File: tags/3.3.34/FarNetPlugin/InputBox.cpp
/*
Far.NET plugin for Far Manager
Copyright (c) 2005-2007 Far.NET Team
*/

#include "StdAfx.h"
#include "InputBox.h"

namespace FarManagerImpl
{;
InputBox::InputBox()
{
	_maxLength = 511;
}

int InputBox::MaxLength::get()
{
	return _maxLength;
}

void InputBox::MaxLength::set(int value)
{
	if (value < 1)
		throw gcnew ArgumentOutOfRangeException("value");
	_maxLength = value;
}

String^ InputBox::HelpTopic::get() { return _HelpTopic; }
void InputBox::HelpTopic::set(String^ value)
{
	if (SS(value) && !value->StartsWith("<"))
		throw gcnew ArgumentException("'value' format must be '<FullPath\\>Topic'");
	_HelpTopic = value;
}

bool InputBox::Show()
{
	CStr sTitle(Title);
	CStr sPrompt(Prompt);
	CStr sText(Text);
	CStr sDest(_maxLength + 1);
	CStr sHistory;
	if (SS(History))
		sHistory.Set(History);

	// help
	const char* help = 0; CStr sHelp;
	if (_oemHelpTopic)
	{
		help = _oemHelpTopic;
	}
	else if (SS(HelpTopic))
	{
		sHelp.Set(HelpTopic);
		help = sHelp;
	}

	if (!Info.InputBox(sTitle, sPrompt, sHistory, sText, sDest, MaxLength, help, Flags()))
		return false;

	Text = OemToStr(sDest);
	return true;
}

int InputBox::Flags()
{
	int r = 0;
	if (ButtonsAreVisible) r += FIB_BUTTONS;
	if (EmptyEnabled) r += FIB_ENABLEEMPTY;
	if (EnvExpanded) r += FIB_EXPANDENV;
	if (IsPassword) r += FIB_PASSWORD;
	if (NoLastHistory) r += FIB_NOUSELASTHISTORY;
	return r;
}

}

File: tags/4.3.0/FarNet/FarNet/InputBox.cs
/*
FarNet plugin for Far Manager
Copyright (c) 2005 FarNet Team
*/

namespace FarNet
{
	/// <summary>
	/// Input box. It is created by <see cref="IFar.CreateInputBox"/>.
	/// </summary>
	public interface IInputBox
	{
		/// <summary>
		/// Text to be edited.
		/// </summary>
		string Text { get; set; }
		/// <summary>
		/// Title of the box.
		/// </summary>
		string Title { get; set; }
		/// <summary>
		/// Prompt text.
		/// </summary>
		string Prompt { get; set; }
		/// <summary>
		/// History string.
		/// </summary>
		string History { get; set; }
		/// <summary>
		/// Maximal length of <see cref="Text"/>.
		/// </summary>
		int MaxLength { get; set; }
		/// <summary>
		/// Enabled empty input.
		/// </summary>
		bool EmptyEnabled { get; set; }
		/// <summary>
		/// Tells that it is used for file system path input.
		/// </summary>
		/// <remarks>
		/// Setting this to true enables some extras, e.g. on typing: a dropdown list of matching available paths.
		/// </remarks>
		bool IsPath { get; set; }
		/// <summary>
		/// Tells that it is used for password input.
		/// </summary>
		/// <remarks>
		/// If it is true then asterisks are displaied instead of input characters.
		/// </remarks>
		bool IsPassword { get; set; }
		/// <summary>
		/// Expand environment variables.
		/// </summary>
		bool EnvExpanded { get; set; }
		/// <summary>
		/// If <see cref="Text"/> is empty and <see cref="History"/> is not empty,
		/// then do not initialize the input line from the history.
		/// </summary>
		bool NoLastHistory { get; set; }
		/// <summary>
		/// Buttons are visible.
		/// </summary>
		bool ButtonsAreVisible { get; set; }
		/// <summary>
		/// The only supported format is "&lt;FullPath\&gt;Topic", see <see cref="IFar.ShowHelp"/>.
		/// </summary>
		string HelpTopic { get; set; }
		/// <summary>
		/// Shows input box and waits until user press OK or Cancel.
		/// </summary>
		/// <returns>True if OK was pressed.</returns>
		bool Show();
	}
}